src/rust/engine/options/src/config.rs

```diff
@@ -336,21 +338,46 @@ impl ConfigReader {
         id.name("_", NameTransform::None)
     }
 
-    fn get_value(&self, id: &OptionId) -> Option<&Value> {
+    fn get_from_section(&self, section_name: &str, option_name: &str) -> Option<&Value> {
         self.config
             .value
-            .get(id.scope.name())
-            .and_then(|table| table.get(Self::option_name(id)))
+            .get(section_name)
+            .and_then(|table| table.get(option_name))
+    }
+
+    fn get_value(&self, id: &OptionId) -> Option<&Value> {
+        let option_name = Self::option_name(id);
+        self.get_from_section(id.scope.name(), &option_name)
+            .or(self.get_from_section(DEFAULT_SECTION, &option_name))
     }
 
     fn get_list<T: FromValue + Parseable>(
         &self,
         id: &OptionId,
+    ) -> Result<Option<Vec<ListEdit<T>>>, String> {
+        let from_scoped_section_opt = self.get_list_from_section(id.scope.name(), id)?;
+
+        Ok(
+            if let Some(from_default_section) = self.get_list_from_section(DEFAULT_SECTION, id)? {
+                Some(itertools::concat([
+                    from_default_section,
+                    from_scoped_section_opt.unwrap_or(vec![]),
+                ]))
```

Comment:
Totally minor code style thing (no need to change), but I imagine the nifty `unwrap_or` approach might work for both?

```rust
let from_default_section_opt = self.get_list_from_section(DEFAULT_SECTION, id)?;
let from_scoped_section_opt = self.get_list_from_section(id.scope.name(), id)?;

Ok(Some(itertools::concat([
  from_default_section_opt.unwrap_or(vec![]),
  from_scoped_section_opt.unwrap_or(vec![]),
]))
```

Reply:
Not quite - the current code will return `Ok(None)` if both sections are empty, yours will return `Ok(Some([]))`, which isn't what we want.

Reply:
And note that we do want to return `Ok(Some([]))` if config specifies an empty list explicitly, so this distinction does matter.

Reply:
Ah, yes, I see the change in return value.

> if config specifies an empty list explicitly

Just diving into this to check my understanding of this code: I was under the impression that the config specifying an empty list would come through as `vec![ListEdit { action: Replace, items: vec![]}]`, not `vec![]`. Is that correct? Or am I thinking about something else?

That is, for config like:

```toml
[foo]
opt = []
```

Reply:
Ah yes, an explicit empty list would give that. I was thinking of the case where the edits list is empty. But I guess that can't actually happen due to: https://github.com/pantsbuild/pants/blob/93c272fe14bd00ba0a654282247143c7bc91e763/src/rust/engine/options/src/config.rs#L439 where we coerce that case to None (and I'm not sure that case is possible in TOML anyway, since something has to be on the RHS of the key).

So I guess this might be possible. But some tests failed when I tried it, so 🤷

Reply:
Ah cool, thanks for explaining. (Just confirming no change necessary from my perspective.)